extensions/amp-ad/0.1/amp-ad-3p-impl.js

```diff
@@ -0,0 +1,438 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {IntersectionObserver} from '../../../src/intersection-observer';
+import {getAdCid} from '../../../src/ad-cid';
+import {getIframe, prefetchBootstrap} from '../../../src/3p-frame';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {listenFor, listenForOnce,
+    postMessage} from '../../../src/iframe-helper';
+import {loadPromise} from '../../../src/event-helper';
+import {parseUrl} from '../../../src/url';
+import {adPrefetch, adPreconnect} from '../../../ads/_config';
+import {timer} from '../../../src/timer';
+import {user} from '../../../src/log';
+import {viewerFor} from '../../../src/viewer';
+import {removeElement} from '../../../src/dom';
+
+/** @const These tags are allowed to have fixed positioning */
+const POSITION_FIXED_TAG_WHITELIST = {
+  'AMP-FX-FLYING-CARPET': true,
+  'AMP-LIGHTBOX': true,
+  'AMP-STICKY-AD': true,
+};
+
+/**
+ * Store loading ads info within window to ensure it can be properly stored
+ * across separately compiled binaries that share load throttling.
+ * @const ID of window variable used to track 3p ads waiting to load.
+ */
+const LOADING_ADS_WIN_ID_ = '3pla';
+
+/**
+ * @param {!Element} element
+ * @param {!Window} win
+ * @return {number|boolean}
+ */
+export function allowRenderOutsideViewport(element, win) {
+  // Store in window Object that serves as a set of timers associated with
+  // waiting elements.
+  const loadingAds = win[LOADING_ADS_WIN_ID_] || {};
```

Comment:
We want to protect against the race condition of the timer or the promise chain returning first.  We do not want both of them to decrement the count so using the timerId ensures it is only ever done once.